scripts/sandbox.js

```diff
@@ -1,40 +1,66 @@
-let sandboxLevel = "";
-let sandboxObject = ObstacleCube;
-let sanboxTpX = 3;
-let sandboxTpY = 3;
-let sandboxPortalColor = "Blue";
+const sandbox = {
+  history: [null],
+  level: "",
+  object: ObstacleCube,
+  tpX: 0,
+  tpY: 0,
+  portalColor: "Blue",
+};
 
 class Sandbox {
   static update() {
     UserInterface.displayText({
-      content: `x: ${Math.floor(mouseX / UNIT_SIZE)}, y:${Math.floor(
+      content: `x: ${Math.floor(mouseX / UNIT_SIZE)}, y: ${Math.floor(
         mouseY / UNIT_SIZE
       )}`,
       x: 1,
       y: 2,
     });
     this.drawMarking(mouseX, mouseY);
   }
+
+  static undo() {
+    const history = sandbox.history;
+
+    if (history.length - 1 > 0) {
+      history.splice(history.length - 1, 1);
```

Comment:
It is good practice to not mutate objects in programming, since it may lead to unwanted side effects. You can achieve the same effect here with either of these methods.

```javascript
sandbox.history = history.slice(0, -1)
sandbox.history = history.slice(0, history.length - 1)
```

Mutating an object btw, is when you change it's properties without replacing it.

```javascript
let mutatedArray = [1, 2, 3]
mutatedArray.push(4)
console.log(mutatedArray)
//=> [1, 2, 3, 4]

const nonMutatedArray = [1, 2, 3]
const newArray  = nonMutatedArray.concat(4)
console.log(nonMutatedArray)
//=> [1, 2, 3]
console.log(newArray)
//=> [1, 2, 3, 4]
```

Reply:
Just to be clear, it seems you are mistaking my `splice()` with `slice()`, those are different things.

Reply:
No I know. In this instance you could use slice :)